### DISCUSSION TITLE
Visualizations with multi-objectives

### DISCUSSION THREAD

**Discussion by DeLaRiva** [optuna/optuna#3825] [Q&A] [answered]
I'm implementing multi objectives and would like to get some visualizations afterwards. So far I have:

```
sampler = optuna.integration.BoTorchSampler(
                                                    # constraints_func=constraints,
                                                    n_startup_trials = TRIALS
                                                    )

study = optuna.create_study(
                                              directions = ["maximize", "minimize"]
                                              , sampler = sampler
                                             )

study.optimize(
                        lambda trial: objective(trial, tube_name)
                        , n_trials = n_trials
                        , show_progress_bar=True
                        )

fig = optuna.visualization.plot_pareto_front(study, target_names=["Dose", "filament"])
fig.show()

fig = optuna.visualization.plot_contour(study, target_name = ["Dose"])
fig.show()
```

As far as I can tell it works except of the second plot. For this I'm receiving following:

```
Traceback (most recent call last):
  File "C:\Users\User\project\ViPar.py", line 227, in <module>
    fig = optuna.visualization.plot_contour(study, target_name = ["Dose"])
  File "C:\Users\User\Anaconda3\lib\site-packages\optuna\visualization\_contour.py", line 89, in plot_contour
    _check_plot_args(study, target, target_name)
  File "C:\Users\User\Anaconda3\lib\site-packages\optuna\visualization\_utils.py", line 48, in _check_plot_args
    raise ValueError(
ValueError: If the `study` is being used for multi-objective optimization, please specify the `target`.
```
I wonder about the reason for this issue. Don't I specify the target in visualization.plot_contour() ? The Pareto plot is working.
As I just realized while typing that I forgot target() which is needed when using multi objectives, I included this now as following:

`fig = optuna.visualization.plot_contour(study, target = "Dose", target_name = ["Dose"])`

which is not working as well with this response:

```
Traceback (most recent call last):
  File "C:\Users\DWIEBWohlfahrt\OneDrive - Smiths Group\Desktop\Datenanalyse\Thors_Datenhammer\Betatron\ViPar.py", line 227, in <module>
    fig = optuna.visualization.plot_contour(study, target = "Dose", target_name = ["Dose"])
  File "C:\Users\User\Anaconda3\lib\site-packages\optuna\visualization\_contour.py", line 90, in plot_contour
    return _get_contour_plot(study, params, target, target_name)
  File "C:\Users\User\Anaconda3\lib\site-packages\optuna\visualization\_contour.py", line 189, in _get_contour_plot
    sub_plots = _generate_contour_subplot(
  File "C:\Users\User\Anaconda3\lib\site-packages\optuna\visualization\_contour.py", line 281, in _generate_contour_subplot
    value = target(trial)
TypeError: 'str' object is not callable
```

The [documentation ](https://optuna.readthedocs.io/en/latest/reference/visualization/generated/optuna.visualization.plot_contour.html#optuna.visualization.plot_contour) says 
```
target ([Optional](https://docs.python.org/3/library/typing.html#typing.Optional)[[Callable](https://docs.python.org/3/library/typing.html#typing.Callable)[[[FrozenTrial](https://optuna.readthedocs.io/en/latest/reference/generated/optuna.trial.FrozenTrial.html#optuna.trial.FrozenTrial)], [float](https://docs.python.org/3/library/functions.html#float)]]) –

A function to specify the value to display. If it is [None](https://docs.python.org/3/library/constants.html#None) and study is being used for single-objective optimization, the objective values are plotted.

Note: Specify this argument if study is being used for multi-objective optimization.
```

what function shall I specify? I would like to have the exact plots like when performing a single objective. Is(n't) this possible?

Thanks a lot in advance!

**Comment by nzw0301** [comment, marked as answer]
As in the docs, `target` takes a trial instance and returns a float value, for example, `target=lambda t: t.values[0]` when we use the first dimension of the multi-objective values for the visualisation.

**Reply by DeLaRiva** [reply]
Thank you. I've seen this but I can't get it to work.

```
        study.optimize(
                        lambda trial: objective(trial, tube_name)
                        , n_trials = n_trials
                        , show_progress_bar=True
                        )

        fig = optuna.visualization.plot_pareto_front(study, target_names=["Dose", "filament"])
        fig.show()

        trial_with_highest_dose = max(study.best_trials, key=lambda t: t.values[1])
        print("\n")
        print(f"\tnumber: {trial_with_highest_dose.number}")
        print(f"\tparams: {trial_with_highest_dose.params}")
        print(f"\tvalues: {trial_with_highest_dose.values}")
        print("\n")
        fig = optuna.visualization.plot_contour(study, target = lambda t: t.values[0], target_name = ["Dose"])
        fig.show()
```

Everything works except of the plot_contour():

> Traceback (most recent call last):
>   File "C:\Users\User\project\ViPar.py", line 230, in <module>
>     fig = optuna.visualization.plot_contour(study, target = lambda t: t.values[0], target_name = ["Dose"])
>   File "C:\Users\User\Anaconda3\lib\site-packages\optuna\visualization\_contour.py", line 90, in plot_contour
>     return _get_contour_plot(study, params, target, target_name)
>   File "C:\Users\User\Anaconda3\lib\site-packages\optuna\visualization\_contour.py", line 189, in _get_contour_plot
>     sub_plots = _generate_contour_subplot(
>   File "C:\Users\User\Anaconda3\lib\site-packages\optuna\visualization\_contour.py", line 291, in _generate_contour_subplot
>     contour = go.Contour(
>   File "C:\Users\User\Anaconda3\lib\site-packages\plotly\graph_objs\_contour.py", line 3049, in __init__
>     self["colorbar"] = _v
>   File "C:\Users\User\Anaconda3\lib\site-packages\plotly\basedatatypes.py", line 4851, in __setitem__
>     self._set_compound_prop(prop, value)
>   File "C:\Users\User\Anaconda3\lib\site-packages\plotly\basedatatypes.py", line 5262, in _set_compound_prop
>     val = validator.validate_coerce(val, skip_invalid=self._skip_invalid)
>   File "C:\Users\User\Anaconda3\lib\site-packages\_plotly_utils\basevalidators.py", line 2455, in validate_coerce
>     v = self.data_class(v, skip_invalid=skip_invalid, _validate=_validate)
>   File "C:\Users\User\Anaconda3\lib\site-packages\plotly\graph_objs\contour\_colorbar.py", line 2125, in __init__
>     self["title"] = _v
>   File "C:\Users\User\Anaconda3\lib\site-packages\plotly\basedatatypes.py", line 4851, in __setitem__
>     self._set_compound_prop(prop, value)
>   File "C:\Users\User\Anaconda3\lib\site-packages\plotly\basedatatypes.py", line 5262, in _set_compound_prop
>     val = validator.validate_coerce(val, skip_invalid=self._skip_invalid)
>   File "C:\Users\User\Anaconda3\lib\site-packages\_plotly_utils\basevalidators.py", line 2488, in validate_coerce
>     return super(TitleValidator, self).validate_coerce(v, skip_invalid=skip_invalid)
>   File "C:\Users\User\Anaconda3\lib\site-packages\_plotly_utils\basevalidators.py", line 2464, in validate_coerce
>     self.raise_invalid_val(v)
>   File "C:\Users\User\Anaconda3\lib\site-packages\_plotly_utils\basevalidators.py", line 288, in raise_invalid_val
>     raise ValueError(
> ValueError:
>     Invalid value of type 'builtins.list' received for the 'title' property of contour.colorbar
>         Received value: ['Dose']
> 
>     The 'title' property is an instance of Title
>     that may be specified as:
>       - An instance of :class:`plotly.graph_objs.contour.colorbar.Title`
>       - A dict of string/value properties that will be passed
>         to the Title constructor
> 
>         Supported dict properties:
> 
>             font
>                 Sets this color bar's title font. Note that the
>                 title's font used to be set by the now
>                 deprecated `titlefont` attribute.
>             side
>                 Determines the location of color bar's title
>                 with respect to the color bar. Defaults to
>                 "top" when `orientation` if "v" and  defaults
>                 to "right" when `orientation` if "h". Note that
>                 the title's location used to be set by the now
>                 deprecated `titleside` attribute.
>             text
>                 Sets the title of the color bar. Note that
>                 before the existence of `title.text`, the
>                 title's contents used to be defined as the
>                 `title` attribute itself. This behavior has
>                 been deprecated.

Almost the same syntax works fine for the highest parameters: `trial_with_highest_dose = max(study.best_trials, key=lambda t: t.values[1])`

What is it I am doing wrong?

**Reply by DeLaRiva** [reply]
Ok, I found it..  it was written in the terminal response and I saw it but actually I didn't..
the mistake I made was, that I wrote ["Dose"] instead of "Dose" for the target_name.

Thanks nevertheless!
